tools/eventcontextcodegen/context_parser.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"reflect"
+	"strings"
+)
+
+type targetContextField struct {
+	fieldType fieldType
+	name      string
+	source    string
+}
+
+type targetContextDefinition struct {
+	fields []targetContextField
+}
+
+func parseEventContextFile(filePath string, src string) (targetContextDefinition, error) {
+	// Parse the Go source file
+	fset := token.NewFileSet() // positions are relative to fset
+	node, err := parser.ParseFile(fset, filePath, src, parser.ParseComments)
+	if err != nil {
+		return targetContextDefinition{}, fmt.Errorf("cannot parse file: %w", err)
+	}
+	for _, decl := range node.Decls {
+		structType, ok := isDefinition(decl, "EventContext")
+		if !ok {
+			continue
+		}
+
+		return parseTargetEventContextDefinition(structType)
+	}
+
+	return targetContextDefinition{}, errors.New("could not find target event context definition!")
+}
+
+func parseTargetEventContextDefinition(n *ast.StructType) (targetContextDefinition, error) {
+	if n.Fields.NumFields() == 0 {
+		return targetContextDefinition{}, nil
+	}
+
+	fields, err := parseTargetContextStructFields(nil, n)
+	if err != nil {
+		return targetContextDefinition{}, err
+	}
+
+	return targetContextDefinition{
+		fields: fields,
+	}, nil
+}
+
+func parseTargetContextStructFields(path []string, n *ast.StructType) ([]targetContextField, error) {
+	var result []targetContextField
+
+	for i, f := range n.Fields.List {
+		switch field := f.Type.(type) {
+		case *ast.StructType:
+			fmt.Printf("unsupported struct type for field no %d in target\n", i)
+			continue
+
+		case *ast.ArrayType:
+			elemType, ok := field.Elt.(*ast.Ident)
+			if !ok {
+				return nil, fmt.Errorf("unsupported array element type: %s", reflect.TypeOf(elemType).String())
```

Comment:
`%T` fmt verb prints the underlying type, so you don't need to call reflect yourself.

Reply:
Ahh LOL I was using `%t` and was confused why it didn't print correctly. Nice catch 👌